pkg/reconcile/pipeline/context/impl_test.go

```diff
@@ -91,7 +94,182 @@ var _ = Describe("Context", func() {
+
+			ctx := &impl{client: client, serviceBinding: &sb, typeLookup: typeLookup}
+
+			applications, err := ctx.Applications()
+			Expect(err).NotTo(HaveOccurred())
+			Expect(applications).To(HaveLen(2))
+			Expect(applications[0].Resource()).To(Equal(u1))
+			Expect(applications[0].ContainersPath()).To(Equal(expectedContainerPath))
+			Expect(applications[1].Resource()).To(Equal(u2))
+			Expect(applications[1].ContainersPath()).To(Equal(expectedContainerPath))
+		},
+			Entry("no binding path specified", nil, defaultContainerPath),
+			Entry("binding path specified", &v1alpha1.BindingPath{ContainersPath: "foo.bar"}, "foo.bar"),
+		)
+		DescribeTable("should return slice of size 0 if no application is matching through label seclector", func(bindingPath *v1alpha1.BindingPath, expectedContainerPath string) {
+			ls := &metav1.LabelSelector{
+				MatchLabels: map[string]string{"env": "prod"},
+			}
+
+			ref := v1alpha1.Application{
+				Ref: v1alpha1.Ref{
+					Group:   "app",
+					Version: "v1",
+					Kind:    "Foo",
+				},
+				LabelSelector: ls,
+				BindingPath:   bindingPath,
+			}
+
+			sb := v1alpha1.ServiceBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "sb1",
+					Namespace: "ns1",
+				},
+				Spec: v1alpha1.ServiceBindingSpec{
+					Application: &ref,
+				},
+			}
+			gvr := &schema.GroupVersionResource{Group: "app", Version: "v1", Resource: "foos"}
+			typeLookup.EXPECT().ResourceForReferable(&ref).Return(gvr, nil)
+
+			u := &unstructured.Unstructured{}
+			u.SetName("app")
+			u.SetNamespace(sb.Namespace)
+			u.SetGroupVersionKind(schema.GroupVersionKind{Group: "app", Version: "v1", Kind: "Foo"})
+			u.SetLabels(map[string]string{"env": "stage"})
+			client := fake.NewSimpleDynamicClient(runtime.NewScheme(), u)
+
+			ctx := &impl{client: client, serviceBinding: &sb, typeLookup: typeLookup}
+
+			applications, err := ctx.Applications()
+			Expect(err).NotTo(HaveOccurred())
+			Expect(applications).To(HaveLen(0))
+		},
+			Entry("no binding path specified", nil, defaultContainerPath),
+			Entry("binding path specified", &v1alpha1.BindingPath{ContainersPath: "foo.bar"}, "foo.bar"),
+		)
+		It("should return error if both application name and label is specified", func() {
+			ls := &metav1.LabelSelector{
+				MatchLabels: map[string]string{"env": "prod"},
+			}
+
+			ref := v1alpha1.Application{
+				Ref: v1alpha1.Ref{
+					Group:   "app",
+					Version: "v1",
+					Kind:    "Foo",
+					Name:    "app1",
+				},
+				LabelSelector: ls,
+			}
+
+			sb := v1alpha1.ServiceBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "sb1",
+					Namespace: "ns1",
+				},
+				Spec: v1alpha1.ServiceBindingSpec{
+					Application: &ref,
+				},
+			}
+
+			u := &unstructured.Unstructured{}
+			u.SetName("app")
+			u.SetNamespace(sb.Namespace)
+			u.SetGroupVersionKind(schema.GroupVersionKind{Group: "app", Version: "v1", Kind: "Foo"})
+			u.SetLabels(map[string]string{"env": "stage"})
```

Comment:
we do not need to create any app object at all, since we are going to fail earlier. Omitting this will also implicitly check that we do not try to access the cluster for this scenario.

Reply:
This code has been removed. Ref.  #974